### PR TITLE
#14 initial emoji keywords

### PR DESCRIPTION
**Pull request by wkyoshida**
**Description**
- Added processing to get an initial _'keyword-to-emoji'_ mapping

**Some notes:**
- Realized that the [`cldr-annotations-derived-full`](https://github.com/unicode-org/cldr-json/tree/main/cldr-json/cldr-annotations-derived-full/annotationsDerived) package appears to contain, for the most part, emoji variants actually. I believe hence the _derived_ descriptor. These are variants mostly, from what I could tell, for different skin-tone options, for instance, on some emojis. Geographical flags also appear to be derived variants as well. 
  - **Next steps** for this I'm thinking will be mostly how to deal with skin-tone variants. What might make sense is to find the keywords for the base emoji - the keywords would apply equally for the base and all variants of the emoji. The difference here is that Scribe would need a way to keep track of the skin-tone preference of the user and then auto-suggest the correct variants.  
- Added the [`cldr-annotations-full`](https://github.com/unicode-org/cldr-json/tree/main/cldr-json/cldr-annotations-full/annotations) package, since this one is the package that appears to have a lot of the common emojis as well.  
  - Found that there are also many non-emoji characters that have annotations though. Thinking of not including them.
  - To do the above, actually decided to bring in [emoji](https://github.com/carpedm20/emoji/) :laughing: since it has a [`EMOJI_DATA`](https://carpedm20.github.io/emoji/docs/api.html#emoji.EMOJI_DATA) dict that I thought could be used to check which characters in the CLDR package are indeed emojis.
- For now, decided to only use annotations that are a singular word, since they can more reliably be used as a trigger keyword. Annotations that are phrases, for instance `"grinning face with big eyes",` for `"😃"`, opens up too many different ways of having to parse keywords out for now. Maybe in the future perhaps.  
- **Next steps** in general will be to improve on the keyword generation:
  - Thinking through what to do with keywords that are too general. For instance, `"face"` is a keyword that maps to several smiley emojis. Will have to look into how to determine the top 3 emojis to use for the auto-suggest. Perhaps popularity, relevance (somehow), or simply whatever 3 came first.
  - Bringing in [nltk](https://github.com/nltk/nltk) for stemming as already discussed.

### CODE REVIEW

**Comment by andrewtavis** [comment]
Thank you, @wkyoshida!! Really looking forward to diving into this and testing what's here 😊 Will hopefully get to this in the next few days :)

**Comment by wkyoshida** [comment]
Hey @andrewtavis, I added some changes to hopefully help with some testing and visualizing of the current processing.
- [cbf36c8](https://github.com/scribe-org/Scribe-Data/pull/25/commits/cbf36c8be1c322e10e1aa798d2444a8d1472749c): I was previously testing things out by running the `process_unicode.py` file directly. So added a step to the Jupyter notebook so the code can be run from there instead. 
-  [405c290](https://github.com/scribe-org/Scribe-Data/pull/25/commits/405c290f73cdc81cb544bc4fff4feb8b03121b52): I was already saving the generated files temporarily for myself just to see what they look like, but I didn't commit that code since I still wanted to further iron out before some of the **Next steps** mentioned in the PR description. I added the change here so the generated files get saved locally still though; hopefully this helps visualizing what the data looks like so far.
    - I will end up cleaning up this change once work on the actual code gets worked on, but feel free to revert this particular commit if we'd like to keep the `main` branch cleaner.

**Comment by andrewtavis** [comment]
@wkyoshida, just FYI I’m a bit under the weather, hence no action on this yet 🤒 Wanted to reach out to let you know and reiterate that I’m looking forward to checking all this out 😊

**Comment by wkyoshida** [comment]
> @wkyoshida, just FYI I’m a bit under the weather, hence no action on this yet :face_with_thermometer: Wanted to reach out to let you know and reiterate that I’m looking forward to checking all this out :blush:

Ahh that sucks! Sorry to hear, @andrewtavis.. Hope you get well soon!! :muscle:
No worries at all, and take your time to take care of yourself!

I still plan to steadily keep chipping away on the emoji work :+1: 
I just added [a3333be](https://github.com/scribe-org/Scribe-Data/pull/25/commits/a3333be7914b2a100d4a19a1c382838f1dbc7be4) so that both `cldr-annotations-derived-full` and `cldr-annotations-full` data are getting processed. Still have to implement a way to only consider the base emoji for the case of the derived emoji variants though. Will look into that tomorrow likely.

**Comment by andrewtavis** [comment]
It's looking good so far, @wkyoshida! Do you want to do this all within one PR with partial reviews then? What steps do you think are remaining?

Thanks for your efforts on this! Really great to have someone contributing on the data end of things 🚀

**Comment by wkyoshida** [comment]
Of course - glad to help! :+1: 

Oh feel free to merge this PR if you'd like! That's no problem. I'm fine with opening up new PRs as work gets done on this. I've been pushing smaller chunks at a time, more so, so that it's easier to see the direction I'm going and we can continue to discuss details as things come up. The `Next steps` I laid out in the PR description are definitely what I'm looking to tackle, well.., next :laughing: but they don't have to all be in the same PR.

**Comment by andrewtavis** [comment]
Ok, wonderful :) Will look to merge what you have tomorrow evening and then we can work on other PRs. Let me know if you have any questions 😊 

Also thanks for laying out the plan for this as well as you have. To be expected :) I only tangentially looked at the commit messages, but see now that it’s all there 🙏

Hope you’re having a nice weekend!

**Comment by wkyoshida** [comment]
> Hope you’re having a nice weekend!

I am, thank you! It seems that you've been more back as well; so hopefully you're already feeling better! :muscle::pill: 

---

**Some updates:**
- I added [6502fa8](https://github.com/scribe-org/Scribe-Data/pull/25/commits/6502fa8fe50dae3760a89a34338d9a6d5e00174c) because I saw a note in www.unicode.org/reports/tr51/#Emoji_Implementation_Notes that only fully-qualified emoji should be generated by keyboards. I'm thinking that perhaps that applies for Scribe's case. This commit filters for that now.
- I didn't get to fully implementing a way to only consider the base emoji for the case of the derived emoji variants, but some ideas could be to identify the base emoji by looking at the `Emoji_Modifier_Base` property ([unicode docs](http://www.unicode.org/reports/tr51/#Emoji_Properties)). It looks like there might be a way to do it via pyciu, but still looking into it.
- Also on the base emoji topic, I am wondering how to store them as well. I've thought of a couple ways:
  - Add some field in the JSON file to mark the emojis as a base or non-base emoji
    - As the user types, for example, keyboard could then just check if field `is_base` is `True`  
  - Keep a separate file just to list the base emojis
    - As the user types, keyboard could just check if the potential suggestion exists in this base emoji list
  - Keep two JSON files, one for regular emojis and one for the base emojis
    - As the user types, keyboard could just check for suggestions in both JSON files
  - Not store anything additional
    - As the user types, keyboard could just check every emoji suggestion if it is a base emoji or not

  I think I'm leaning on the first for now, but let me know if there are any firm thoughts on this.    
- Regarding the `Next step` on identifying which suggestions to use for keywords that are too general, e.g. `"face"`, fortunately, it does look like there is [emoji frequency data](https://home.unicode.org/emoji/emoji-frequency/) available. There are also other [data resources](https://emojipedia.org/stats/). So it does look like the top emoji suggestions could be determined via popularity :smile: `TODO` on looking into how to best incorporate this data

**Comment by andrewtavis** [comment]
Am feeling a bit better, @wkyoshida, but am also catching up with work atm. I'll get to all of this tomorrow 😊 Let me just say that in skimming all this it's all really great! Very impressed by the detail you're putting into everything that's also above and beyond research fueled :) Sorry that I've been late in getting to all this, but it will definitely be a priority for tomorrow!

**Comment by wkyoshida** [comment]
No worries, @andrewtavis! And thanks! :smiley: 

With [8e3f8ef](https://github.com/scribe-org/Scribe-Data/pull/25/commits/8e3f8ef20186f5209b7ae63e9043c604100adc06), I believe I figured out a way to process for the base emoji for variants. Using pyicu, emojis can be checked if they fall under the [`EMOJI_MODIFIER_BASE`](https://unicode-org.github.io/icu-docs/apidoc/released/icu4c/uchar_8h.html#ae40d616419e74ecc7c80a9febab03199acd0dcc4831878bf6e74b6257120185d5) property. Since variants are a combination of multiple code points. For instance, `🤟🏾` is really the combination of: `U+1F91F` `U+1F3FE` or `🤟` `🏾`, I'm thinking that a check for emojis that are a single code point and have the property could suffice. Perhaps this logic could be improved over time, but from what I was able to see in generated results, it seemed to satisfy.  

Also, the script is saving now if the emoji is a base emoji or not via a `is_base` field. So each emoji listed looks something like: 
```json
{
"emoji": "🤟",
"is_base": true
},
```

Regarding [81be78c](https://github.com/scribe-org/Scribe-Data/pull/25/commits/81be78c608c829435b9d0009ea3265f5e21faf17), PyICU does have some dependencies that can't simply be resolved via `pip` though.  I was looking into perhaps leveraging a `Brewfile` to resolve for those dependencies (since `brew` can be used for both MacOS and Linux), but I eventually landed on simply noting down to make sure to reference PyICU's documentation. Fortunately, the dependencies seem pretty standard (I already had them in my local device anyways). Let me know if there are any thoughts on this. The `Brewfile` would have something like: 
```sh
brew "icu4c"
brew "pkg-config"
```

**Comment by andrewtavis** [comment]
@wkyoshida, I'm gonna eat dinner and then check this all out 😊 Random question that I have for you, what would we name this if it wasn't Scribe-Data? I figure that we're making something that could generally benefit a greater audience if we change the update script to take an argument for the output directory, so maybe it makes sense to switch the name over? `language-pack`?

**Comment by wkyoshida** [comment]
> Edit: I have a sick logo for LanguagePacks, just btw :man_artist::wink:

:laughing: :raised_hands: !!

Y'know, this brings out some interesting discussions/considerations. Glad you brought it up! Cause I was thinking through some of this as I dug deeper into Scribe-Data :laughing: I didn't realize until this week that it's actually available on pypi! Stumbled upon it cause I just randomly searched for "scribe" on pypi (it seems I obviously also missed the pypi badge in the `README.md` too :facepalm:). Anyways, I think some interesting considerations regarding the future of Scribe-Data could be:
- Like you mentioned, it could be reworked to be more reusable for general applications. A big thing here would be to restructure it away from Scribe-iOS specifically (and other repos) and instead to any output location.
- However, a counterpoint to the above idea could be simply that, if Scribe-Data is being made available in some form, would it make sense to just make the data itself available as the packages? That could suffice for those interested and could be something akin to the data packs that CLDR has in npm - that is, unless the idea is really more so to make the data-generation tooling available as the package instead.
- Thinking also about the future feature for the downloadable language packs, and even further beyond to that other potential Toolforge idea, some considerations could be:
  - If Scribe-Data packages are just the data itself, they could just be exactly what the keyboards could download
  - If they are the data-generation tooling, then the Toolforge idea could use this tooling as part of its data management/processing

tl;dr I think what becomes of Scribe-Data could go different directions, which will depend on what Scribe would actually want to provide to the community with it.

Now back to your question though :laughing:, honestly, I don't think I'd be opposed to still having "scribe" in the name actually. It would be simply an identifier associating its origin back to the "brand" per se or the team maintaining it. Another idea could be `scribe-languages` or `scribe-language-data` :shrug: just ideas 
Between `LanguageData` and `LanguagePacks` specifically though, I'm thinking perhaps the latter makes more sense. More so, to avoid any confusion with the existing `language-data`

**Comment by wkyoshida** [comment]
Looked a little more into the emoji popularity data, and I think I'm leaning on using the data from [Unicode](https://home.unicode.org/emoji/emoji-frequency/), since:
- It has data for 2021 - so not _today_ _today_ recent, but still fairly recent enough. Not sure if a study for 2022 will come out as well, but that could be used instead too.
- There are some studies as well with data from Google, Twitter, and Apple, but hoping that the Unicode data is more general and not tied to emoji usage in a specific platform.

Disadvantage is that data for emoji popularity doesn't seem to be commonly available for programmatic use, in general. Unicode has a link to this Google Drive [folder](https://drive.google.com/drive/folders/18u0N7OOjbP-V2YolddTPaMojhNFavmmF) with the data - so not ideal, but okay-ish. My guess is that Unicode might publish emoji popularity data every year or two, which could be fine for reference, i.e. I don't think the popularity data that Scribe will use has to be on a _what is it today?_ basis. With that said, since the data that would be used won't be changing often, do you think simply saving the file for the most recent study in the repo could make sense (it would be `2021_ranked.tsv` in the Drive)? It's not an ideal solution, but I figured it's alright enough. [Emojipedia](https://emojipedia.org/)'s emojitracker does appear to have a planned [REST API](https://github.com/emojitracker/emojitrack-rest-api) that could be used, but I don't believe it's currently operational.

**Comment by wkyoshida** [comment]
@andrewtavis I created https://github.com/scribe-org/Scribe-Data/issues/26 to continue talking about Scribe-Data as something that could perhaps be useful for a larger audience. I think there's some interesting potential there! Just need to figure what. That issue can be used for at least the initial discussion/brainstorming :rocket:

**Comment by andrewtavis** [comment]
Hi @wkyoshida! As I stated in #26, very sorry for the absence recently. Lots going on... :) Some comments:

> * Found that there are also many non-emoji characters that have annotations though. Thinking of not including them.

Totally agree with this 👍 Bringing in `emoji` is a nice coincidence in all this 😊

> * As the user types, for example, keyboard could then just check if field `is_base` is `True`

This seems the most sensible to me, as splitting up the files means we'll need to be loading in multiple tables with SQLite. A boolean check in all of this is totally fine :)

> * There are also other [data resources](https://emojipedia.org/stats/). So it does look like the top emoji suggestions could be determined via popularity 😄 `TODO` on looking into how to best incorporate this data

Again, you have my sincere thanks for all your efforts on this 👏 As a baseline I think worst comes to worst we could potentially give earlier emojis precedence over later ones 🤔 No need for it to be perfect from the start.

> I was looking into perhaps leveraging a `Brewfile` to resolve for those dependencies (since `brew` can be used for both MacOS and Linux), but I eventually landed on simply noting down to make sure to reference PyICU's documentation. Fortunately, the dependencies seem pretty standard (I already had them in my local device anyways). Let me know if there are any thoughts on this. The `Brewfile` would have something like:

Python dependencies... 😅 100% ok to do a brew file for this. We can also just get back to this when the next person joins in that has problems getting the env up and running if it's mostly standard.

> With that said, since the data that would be used won't be changing often, do you think simply saving the file for the most recent study in the repo could make sense (it would be `2021_ranked.tsv` in the Drive)?

This makes total sense to me :) :) That file doubtless isn't large enough to worry about size considerations, so we can just save it locally. Seems to be the best option for now, and we can look into [Emojipedia](https://emojipedia.org/)'s later if need be.

**Review by andrewtavis — APPROVED** [review]
I'll do some minor edits to this, @wkyoshida (that I can think of, just to edit the comments so that there's a period at the end as I tend to do full stops for comments that are sentences on their own lines and not for inline comments that I further don't capitalize). 

This really is exactly what we want :) Thanks so much for all your hard work! I really feel bad that the positive feedback that this deserves came late, as the combination of initiative and effort on this fantastic :)
